cartographer_ros/cartographer_ros/assets_writer_main.cc

```diff
@@ -0,0 +1,191 @@
+#include "cartographer/transform/transform_interpolation_buffer.h"
+#include "cartographer_ros/msg_conversion.h"
+#include "cartographer_ros/time_conversion.h"
+#include "gflags/gflags.h"
+#include "glog/logging.h"
+#include "ros/ros.h"
+#include "ros/time.h"
+#include "rosbag/bag.h"
+#include "rosbag/view.h"
+#include "tf2_eigen/tf2_eigen.h"
+#include "tf2_ros/buffer.h"
+#include "urdf/model.h"
+
+DEFINE_string(configuration_directory, "",
+              "First directory in which configuration files are searched, "
+              "second is always the Cartographer installation to allow "
+              "including files from there.");
+DEFINE_string(configuration_basename, "",
+              "Basename, i.e. not containing any directory prefix, of the "
+              "configuration file.");
+DEFINE_string(
+    urdf, "",
+    "URDF file that contains static links for your sensor configuration.");
+DEFINE_string(bag_filename, "", "The bag to run over.");
+DEFINE_string(trajectory_filename, "", "The proto containing the trajectory.
+              Written by the node on /finish_trajectory.");
+
+namespace cartographer_ros {
+  namespace {
+
+  namespace carto = ::cartographer;
+
+  std::unique_ptr<carto::transform::TransformInterpolationBuffer>
+  ReadTrajectory(const std::string& trajectory_filename) {
+    std::ifstream stream(trajectory_filename.c_str());
+    carto::proto::Trajectory trajectory_proto;
+    CHECK(trajectory_proto.ParseFromIstream(&stream));
+    return carto::transform::TransformInterpolationBuffer::FromTrajectory(
+        trajectory_proto);
+  }
+
+  void ReadStaticTransformsFromUrdf(const string& urdf_filename,
+                                    ::tf2_ros::Buffer* buffer) {
+    urdf::Model model;
+    CHECK(model.initFile(urdf_filename));
+    std::vector<boost::shared_ptr<urdf::Link>> links;
+    model.getLinks(links);
+    for (const auto& link : links) {
+      if (!link->getParent() ||
+          link->parent_joint->type != urdf::Joint::FIXED) {
+        continue;
+      }
+
+      const urdf::Pose& pose =
+          link->parent_joint->parent_to_joint_origin_transform;
+      geometry_msgs::TransformStamped transform;
+      transform.transform = ToGeometryMsgTransform(carto::transform::Rigid3d(
+          Eigen::Vector3d(pose.position.x, pose.position.y, pose.position.z),
+          Eigen::Quaterniond(pose.rotation.w, pose.rotation.x, pose.rotation.y,
+                             pose.rotation.z)));
+      transform.child_frame_id = link->name;
+      transform.header.frame_id = link->getParent()->name;
+      buffer->setTransform(transform, "urdf", true /* is_static */);
+    }
+  }
+
+  void Run(const string& trajectory_filename, const string& bag_filename,
+           const string& configuration_directory,
+           const string& configuration_basename, const string& urdf_filename) {
+    ::tf2_ros::Buffer buffer;
+    ReadStaticTransformsFromUrdf(urdf_filename, &buffer);
```

Comment:
Should there be a TODO about handling sensors that are attached via non-fixed joints?